src/lib/charm/openstack/barbican.py

```diff
@@ -71,18 +108,59 @@ def __init__(self):
     @property
     def barbican_api_keystone_pipeline(self):
         if self.keystone_api_version == "2":
-            return 'keystone_authtoken context apiapp'
+            return 'cors keystone_authtoken context apiapp'
         else:
-            return 'keystone_v3_authtoken context apiapp'
+            return 'cors keystone_v3_authtoken context apiapp'
 
     @property
     def barbican_api_pipeline(self):
         return {
-            "2": "keystone_authtoken context apiapp",
-            "3": "keystone_v3_authtoken context apiapp",
-            "none": "unauthenticated-context apiapp"
+            "2": "cors keystone_authtoken context apiapp",
+            "3": "cors keystone_v3_authtoken context apiapp",
+            "none": "cors unauthenticated-context apiapp"
         }[self.keystone_api_version]
 
+    @property
+    def barbican_api_keystone_audit_pipeline(self):
+        if self.keystone_api_version == "2":
+            return 'keystone_authtoken context audit apiapp'
+        else:
+            return 'keystone_v3_authtoken context audit apiapp'
+
+
+class HSMAdapter(charms_openstack.adapters.OpenStackRelationAdapter):
+    """Adapt the barbican-hsm-plugin relation for use in rendering the config
+    for Barbican.  Note that the HSM relation is optional, so we have a class
+    variable 'exists' that we can test in the template to see if we should
+    render HSM parameters into the template.
+    """
+
+    interface_type = 'hsm'
+
+    @property
+    def library_path(self):
+        """Provide a library_path property to the template if it exists"""
+        try:
+            return self.relation.plugin_data['library_path']
+        except:
+            return ''
+
+    @property
+    def login(self):
+        """Provide a login property to the template if it exists"""
+        try:
+            return self.relation.plugin_data['login']
+        except:
+            return ''
+
+    @property
+    def slot_id(self):
+        """Provide a slot_id property to the template if it exists"""
+        try:
+            return self.relation.plugin_data['slot_id']
+        except:
+            return ''
+
 
 class BarbicanAdapters(charms_openstack.adapters.OpenStackRelationAdapters):
     """
```

Comment:
I think you need OpenStackAPIRelationAdapters here, so that you get the amqp, db and cluster relations.